File: src/bindings/java/in3/eth1/Log.java
package in3.eth1;

import in3.*;

/**
 * a log entry of a transaction receipt.
 * 
 */

public class Log {

    private JSON data;

    private Log(JSON data) {
        this.data = data;
    }

    protected static Log[] asLogs(Object o) {
        if (o == null)
            return null;
        if (o instanceof Object[]) {
            Object[] a = (Object[]) o;
            Log[] s = new Log[a.length];
            for (int i = 0; i < s.length; i++)
                s[i] = a[i] == null ? null : new Log((JSON) a[i]);
            return s;
        }
        return null;
    }

    protected static Log asLog(Object o) {
        if (o == null)
            return null;
        return new Log((JSON) o);
    }

    /**
     * true when the log was removed, due to a chain reorganization. false if its a
     * valid log.
     */
    public boolean isRemoved() {
        return (Boolean) data.get("removed");
    }

    /**
     * integer of the log index position in the block. null when its pending log.
     */
    public int getLogIndex() {
        return JSON.asInt(data.get("logIndex"));
    }

    /**
     * integer of the transactions index position log was created from. null when
     * its pending log.
     */
    public int gettTansactionIndex() {
        return JSON.asInt(data.get("transactionIndex"));
    }

    /**
     * Hash, 32 Bytes - hash of the transactions this log was created from. null
     * when its pending log.
     */
    public String getTransactionHash() {
        return data.getString("transactionHash");
    }

    /**
     * Hash, 32 Bytes - hash of the block where this log was in. null when its
     * pending. null when its pending log.
     */
    public String getBlockHash() {
        return data.getString("blockHash");
    }

    /**
     * the block number where this log was in. null when its pending. null when its
     * pending log.
     */
    public long getBlockNumber() {
        return data.getLong("blockNumber");
    }

    /**
     * 20 Bytes - address from which this log originated.
     */
    public String getAddress() {
        return data.getString("address");
    }

    /**
     * Array of 0 to 4 32 Bytes DATA of indexed log arguments. (In solidity: The
     * first topic is the hash of the signature of the event (e.g.
     * Deposit(address,bytes32,uint256)), except you declared the event with the
     * anonymous specifier.)
     */
    public String[] getTopics() {
        return data.getStringArray("topics");
    }

}

File: dotnet/In3/Eth1/Log.cs
using System.Numerics;
using System.Text.Json.Serialization;
using In3.Utils;

namespace In3.Eth1
{
    /// <summary>
    /// Logs marks changes of state on the blockchain for events. The <see cref="Log" /> is a data object with information from logs.
    /// </summary>
    public class Log
    {
        /// <summary>
        /// Flags log removal (due to chain reorganization).
        /// </summary>
        /// <value>If <see langword="true"/> log has been removed, <see langword="false"/> its still a valid log.</value>
        [JsonPropertyName("removed")] public bool Removed { get; set; }

        /// <summary>
        /// Index position in the block.
        /// </summary>
        // TODO: Check if this can be null, i believe it would be converted to 0 on the json converter which is misleading.
        [JsonPropertyName("logIndex"), JsonConverter(typeof(IntFromHexConverter))] public int LogIndex { get; set; }

        /// <summary>
        /// index position log was created from.
        /// </summary>
        // TODO: Same of the above.
        [JsonPropertyName("transactionIndex"), JsonConverter(typeof(IntFromHexConverter))] public int TransactionIndex { get; set; }

        /// <summary>
        /// Hash of the transactions this log was created from. null when its pending log.
        /// </summary>
        [JsonPropertyName("transactionHash")] public string TransactionHash { get; set; }

        /// <summary>
        /// Hash of the block this log was in. null when its pending log.
        /// </summary>
        [JsonPropertyName("blockHash")] public string BlockHash { get; set; }

        /// <summary>
        /// Number of the block this log was in.
        /// </summary>
        // TODO: Verify for nulls
        [JsonPropertyName("blockNumber"), JsonConverter(typeof(BigIntegerFromHexConverter))] public BigInteger BlockNumber { get; set; }

        /// <summary>
        /// Address from which this log originated.
        /// </summary>
        [JsonPropertyName("address")] public string Address { get; set; }

        /// <summary>
        /// Array of 0 to 4 32 Bytes DATA of indexed log arguments. (In solidity: The first topic is the hash of the signature of the event (e.g. Deposit(address,bytes32,uint256)), except you declared the event with the anonymous specifier).
        /// </summary>
        [JsonPropertyName("topics")] public string[] Topics { get; set; }

        /// <summary>
        /// Data associated with the log.
        /// </summary>
        [JsonPropertyName("data")] public string Data { get; set; }

        /// <summary>
        /// Address from which this log originated.
        /// </summary>
        [JsonPropertyName("type")] public string Type { get; set; }
    }
}